fabtests/benchmarks/benchmark_shared.c

```diff
@@ -143,58 +177,136 @@ static int bw_rx_comp()
 	return ft_tx(ep, remote_fi_addr, 4, &tx_ctx);
 }
 
+static int bw_rx_comp_any(int *context_id)
+{
+	int j,ret;
+	struct fi_cq_err_entry comp;
+
+	ret = ft_get_rx_comp_any(&comp);
+
+	if (ret)
+		return ret;
+
+	for (j = 0; j < opts.window_size; ++j) {
+		if (comp.op_context == &rx_ctx_arr[j].context) {
+			*context_id = j;
+			return 0;
+		}
+	}
+
+	FT_ERR("mistmatch context! context: %p\n", comp.op_context);
+	return -FI_EINVAL;
+}
+
 int bandwidth(void)
 {
 	int ret, i, j;
+	struct fi_cq_err_entry comp;
 
 	ret = ft_sync();
 	if (ret)
 		return ret;
 
-	/* The loop structured allows for the possibility that the sender
-	 * immediately overruns the receiving side on the first transfer (or
-	 * the entire window). This could result in exercising parts of the
-	 * provider's implementation of FI_RM_ENABLED. For better or worse,
-	 * some MPI-level benchmarks tend to use this type of loop for measuring
-	 * bandwidth.  */
+	/* The loop structure allows for the two types of windows:
+	 *
+	 *    WAIT_ALL_WINDOW: a new TX/RX operation can be submitted only after
+	 *    all operations in the current window finished.
+	 *
+	 *    WAIT_ANY_WINDOW: a new TX/RX operation can be submitted when there
+	 *    operations finished.
+	 *
+	 * Naturally, WAIT_ANY_WINDOW would result higher bandwidth than
+	 * WAIT_ALL_WINDOW.
+	 *
+	 * For better or worse, some MPI-level benchmarks (such as OSU Micro Benchmark)
+	 * tend to use WAIT_ALL_WINDOW for measuring bandwidth.
+	 *
+	 * Meanwhile, some application/benchmarks (such as NCCL and perftest) uses
+	 * WAIT_ANY_WINDOW fore measuring bandwidth.
+	 */
 
 	if (opts.dst_addr) {
 		for (i = j = 0; i < opts.iterations + opts.warmup_iterations; i++) {
 			if (i == opts.warmup_iterations)
 				ft_start();
 
-			if (opts.transfer_size < fi->tx_attr->inject_size)
-				ret = ft_inject(ep, remote_fi_addr, opts.transfer_size);
-			else
+			if (opts.window_type == WAIT_ALL_WINDOW ||
+			    opts.transfer_size < fi->tx_attr->inject_size) {
+				if (opts.transfer_size < fi->tx_attr->inject_size)
+					ret = ft_inject(ep, remote_fi_addr, opts.transfer_size);
+				else
+					ret = ft_post_tx(ep, remote_fi_addr, opts.transfer_size,
+							 NO_CQ_DATA, &tx_ctx_arr[j].context);
+				if (ret)
+					return ret;
+
+				if (++j == opts.window_size) {
+					ret = bw_tx_comp();
+					if (ret)
+						return ret;
+					j = 0;
+				}
+			} else {
+				if (i < opts.window_size) {
```

Comment:
Some comments and/or better variable names would be appreciated. I think I understand what this is doing but it's not terribly clear.